etc/systemd/system-generators/zfs-mount-generator.in

```diff
@@ -168,6 +170,46 @@ process_line() {
       "${dataset}" >/dev/kmsg
   fi
 
+  # Minimal pre-requisites to mount a ZFS dataset
+  wants="zfs-import.target"
+  if [ -n "${p_encroot}" ] &&
+      [ "${p_encroot}" != "-" ] ; then
+    keyloadunit="zfs-load-key-$(systemd-escape "${p_encroot}").service"
+    if [ "${p_encroot}" = "${dataset}" ] ; then
+        pathdep=""
+      if [ "${p_keyloc%%://*}" = "file" ] ; then
+        pathdep="RequiresMountsFor=${p_keyloc#file://}"
+        keyloadcmd="@sbindir@/zfs load-key ${dataset}"
+      elif [ "${p_keyloc}" = "prompt" ] ; then
+  keyloadcmd="sh -c 'systemd-ask-password | @sbindir@/zfs load-key ${dataset}'"
```

Comment:
Is there a white space issue here? (I’m on mobile, so it may just be displaying weird.)

Reply:
What happens if the passphrase is entered wrong? Does this need to be an sh loop of, say, up to 3 times?

Reply:
I dropped the indentation back to let this fit on a single line, within the style guideline's length limitation. I'll change it to anything that's legible, though.

As for the passphrase issue, I think that should be considered a bug/feature request for systemd, where they can solve this issue for everyone, once and for all, instead of having everyone think about doing this, and maintaining it, themselves.

Reply:
systemd has no way of knowing if the passphrase entered was correct. I suppose the protocol could be changed from outputting the password to exec()ing something (in our case `zfs load-key`) and checking its exit status. But for now, we need to work with what exists.

The sh code isn't particularly complicated, but it's probably too much for a one-liner in the unit. Here's a commit (on a systemd-integration branch on rlaager/zfs) which adds a systemd-load-key script and uses it: https://github.com/rlaager/zfs/commit/a894797ab75b027e09a899c0d1df393309ae27ab

I tested the script standalone. I tested that it gets installed to the right directory.

Reply:
I looked into automatically restarting the unit up to three times on failure, but doing this for `oneshot` services is not allowed, [for some reason](https://github.com/systemd/systemd/issues/2582). A [PR](https://github.com/systemd/systemd/pull/7474) to get that changed stalled years ago, so I think we'll need to implement this functionality ourselves.

I adapted your script, and did include it as a one-liner in the service file, in the interest of making downstream maintainer's lives easier (i.e., to avoid needing to ensure another path is available at some point in the boot process).

I also added an error code return after three failed attempts.